realm/realm-library/src/main/java/io/realm/RealmObjectSchema.java

```diff
@@ -204,6 +204,14 @@ public RealmObjectSchema removeField(String fieldName) {
         if (table.getPrimaryKey() == columnIndex) {
             table.setPrimaryKey(null);
         }
+        // When PK does not exist (-2) this cannot happen. Meanwhile a removal target has a smaller
+        // index than the PK index, we can expect an rearrangement.
+        if (columnIndex < table.getPrimaryKey()) {
+            String pkField = getPrimaryKey();
+            table.removeColumn(columnIndex);
+            table.setPrimaryKey(pkField);
```

Comment:
Table is in the internal API, so we can do anything we want there. Also we shouldn't make bad code just because it might be replaced in a few months :)